packages/apollo-server-core/src/plugin/usageReporting/addTraceRequestStats.ts

```diff
@@ -0,0 +1,428 @@
+        //   in this case is non-nullable. At first glance, it might seem like
+        //   we'd want to avoid saying the field is provided null when its type
+        //   is non-nullable because it might be confusing. But this just avoids
+        //   the truth about an ugly edge case in spec. More importantly, with
+        //   this information, we can let users know when their requests are
+        //   doing this, as it may be unintentional on their part.
+        inputFieldIsUsedByRequest({
+          trace,
+          inputObjectTypeName: getNamedType(parentInputType).name,
+          inputFieldName: node.name.value,
+          inputFieldTypeName: inputType.toString(),
+          isPresent: true,
+          isNull: node.value.kind === 'NullValue' || (
+            node.value.kind === 'Variable' &&
+            nullVariableNames.has(node.value.name.value)
+          ),
+          isUndefined: node.value.kind === 'Variable' &&
+            undefinedVariableNames.has(node.value.name.value),
+        });
+      },
+      EnumValue(node: EnumValueNode): void {
+        // The operation has been successfully validated by this stage, so the
+        // non-null assertion here is fine. Look at the Kind.ENUM case in
+        // TypeInfo.enter() in graphql-js for why this works.
+        const inputType = typeInfo.getInputType()!;
+        enumValueIsPresentInRequest({
+          trace,
+          enumTypeName: getNamedType(inputType).name,
+          enumValueName: node.value,
+        });
+      },
+    }));
+  } catch (_) {
+    // At the stage of AS when this is run, variables have not been validated
+    // yet, and accordingly the code that traverses variables may throw. In
+    // those cases, we consider the request itself as invalid. Since the point
+    // of collecting these stats is to understand how schema changes affect
+    // valid operation executions, we collect no stats for invalid requests.
+    trace.perInputTypeStat = Object.create(null);
+    trace.perEnumTypeStat = Object.create(null);
+  }
+}
+
+// This code is adapted from coerceInputValue() in graphql-js. Note that we
+// validate the variables here in addition to collecting stats, as we don't
+// want to collect stats for invalid requests.
+function addTraceInputValueStats({
+  trace,
+  inputValue,
+  inputType,
+}: {
+  trace: Trace;
+  inputValue: any;
+  inputType: GraphQLInputType;
+}): void {
+  if (isNonNullType(inputType)) {
+    if (inputValue !== null) {
+      addTraceInputValueStats({
+        trace,
+        inputValue,
+        inputType: inputType.ofType,
+      })
+      return;
+    }
+    throw Error('Non-null type cannot be provided null value.');
+  }
+
+  // Provided null for nullable type.
+  if (inputValue === null) return;
+
+  if (isListType(inputType)) {
+    const itemType = inputType.ofType;
+    if (isCollection(inputValue)) {
+      const iterator = (itemValue: any) => {
```

Comment:
The reason I use `isCollection` (and `forEach`) from `iterall` is because these are the same functions and library that `graphql-js` uses to determine whether a JS object can be iterated and how to iterate through it if so. My main worry is just making sure we support what `graphql-js` supports here.